contracts/adapters/GmpTransferAdapter.sol

```diff
@@ -0,0 +1,114 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.11;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165Checker.sol";
+import "../../contracts/interfaces/IBridge.sol";
+import "../../contracts/interfaces/IFeeHandler.sol";
+import "../XERC20/interfaces/IXERC20.sol";
+
+contract GmpTransferAdapter is AccessControl {
+    using ERC165Checker for address;
+
+    IBridge public immutable _bridge;
+    bytes32 public immutable _resourceID;
+    address immutable _gmpAddress;
+
+
+    event Withdrawal(address recipient, uint amount);
+    event FundsTransferred(address recipient, uint256 amount);
+
+    error SenderNotAdmin();
+    error InsufficientMsgValueAmount(uint256 amount);
+    error MsgValueLowerThanFee(uint256 amount);
+    error TokenWithdrawalFailed();
+    error InsufficientBalance();
+    error InvalidHandler(address handler);
+    error InvalidOriginAdapter(address adapter);
+    error FailedFundsTransfer();
+
+    error Invalid_MsgNotZero();
+    error FailedRefund();
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert SenderNotAdmin();
+        _;
+    }
+
+    /**
+        @notice This contract requires for transfer that the origin adapter address is the same across all networks.
+        Because of that it should be deployed using multichain deployer or create2.
+     */
+    constructor(IBridge bridge, address newGmpAddress, bytes32 resourceID) {
+        _bridge = bridge;
+        _gmpAddress = newGmpAddress;
+        _resourceID = resourceID;
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function deposit(uint8 destinationDomainID, address recipientAddress, address XERC20Address, uint256 tokenAmount) external payable {
+        address feeHandlerRouter = _bridge._feeHandler();
+        (uint256 fee, ) = IFeeHandler(feeHandlerRouter).calculateFee(
+            address(this),
+            _bridge._domainID(),
+            destinationDomainID,
+            _resourceID,
+            "", // depositData - not parsed
+            ""  // feeData - not parsed
+        );
+
+        if (msg.value < fee) {
```

Comment:
Why do we collect fees twice?
Once on the adapter and once on the bridge.
Seems reasonable if the the fee is incorrect that the deposit will just revert